pkg/pillar/containerd/containerd.go

```diff
@@ -4,371 +4,492 @@
+	ctrdCtx, ctrdCtxWithLease context.Context
+	deleteCtxLease            func(context.Context) error
+	// CtrdClient is a handle to the current containerd client API
+	CtrdClient   *containerd.Client
+	contentStore content.Store
+)
+
+// InitContainerdClient initializes CtrdClient and ctrdCtx
+func InitContainerdClient() error {
+	log.Infof("InitContainerdClient")
+	var err error
+	if ctrdCtx == nil {
+		ctrdCtx = namespaces.WithNamespace(context.Background(), ctrdServicesNamespace)
+	}
+	if CtrdClient == nil {
+		CtrdClient, err = containerd.New(ctrdSocket, containerd.WithDefaultRuntime(containerdRunTime))
+		if err != nil {
+			log.Errorf("InitContainerdClient: could not create containerd client. %v", err.Error())
+			return fmt.Errorf("initContainerdClient: could not create containerd client. %v", err.Error())
+		}
+	}
+	if ctrdCtxWithLease == nil {
+		ctrdCtxWithLease = namespaces.WithNamespace(context.Background(), ctrdServicesNamespace)
+		ctrdCtxWithLease, deleteCtxLease, err = withLease(ctrdCtxWithLease, -1, ctrdCtxLeaseID)
+		if err != nil {
+			err = fmt.Errorf("InitContainerdClient: could not add lease to containerd context. %v", err.Error())
+			log.Errorf(err.Error())
+			return err
 		}
-		log.Errorf("GetSnapshotID read %s failed: %s", filename, err)
 	}
-	snapshotID := filepath.Base(rootpath)
-	log.Infof("GetSnapshotID basename %s from %s", snapshotID, rootpath)
-	return snapshotID
+	if contentStore == nil {
+		contentStore = CtrdClient.ContentStore()
+	}
+
+	if err := verifyCtr(); err != nil {
+		return fmt.Errorf("InitContainerdClient: exception while verifying ctrd client: %s", err.Error())
+	}
+	return nil
 }
 
-// LoadBlobs load multiple blobs and reference via an image name
-func LoadBlobs(blobs []*types.BlobStatus, ref string) error {
-	// load the content into the containerd content store
-	log.Infof("loadBlobs(%s)", ref)
+//CloseClient closes containerd client
+func CloseClient() error {
+	if err := verifyCtr(); err != nil {
+		return fmt.Errorf("CloseClient: exception while verifying ctrd client: %s", err.Error())
+	}
+	if err := CtrdClient.Close(); err != nil {
+		err = fmt.Errorf("CloseClient: exception while closing containerd client. %v", err.Error())
+		log.Errorf(err.Error())
+		return err
+	}
+	return nil
+}
 
-	var (
-		root           *types.BlobStatus
-		rootDescriptor ocispec.Descriptor
-		index          *ocispec.Index
-		indexHash      string
-		manifests      = make([]*ocispec.Manifest, 0)
-		manifestHashes = make([]string, 0)
-	)
+//CtrWriteBlob reads the blob as raw data from `reader` and writes it into containerd with the given lease
+func CtrWriteBlob(blobHash string, reader io.Reader) error {
+	if err := verifyCtr(); err != nil {
+		return fmt.Errorf("CtrWriteBlob: exception while verifying ctrd client: %s", err.Error())
+	}
+	expectedDigest := digest.Digest(blobHash)
+	if err := expectedDigest.Validate(); err != nil {
+		return fmt.Errorf("CtrWriteBlob: validation failed for digest %s. %v", blobHash, err)
```

Comment:
This message is a bit confusing. This doesn't validate the content, just the hash itself, i.e. is `blobHash` a valid potential hash? If it is `ggt6` it is not, because `g` and `t` are not valid hex chars, and because it is too short for sha256.

Reply:
Changed.

Reply:
Much easier to understand. Thanks